[5.1] Added reservation id to IronMQ delete method

ryantology opened the pull request:
This is required in order to support the v3 API.
IronIO is rapidly moving to the v3 API as it has increased performance.
References Issue #10828 

Due to the incompatibility this likely requires forcing an upgrade of the composer package the is required. Also, there are different endpoints that are defaulted for the new API
Old Endpoint: `mq-aws-us-east-1.iron.io`
New Endpoint: `mq-aws-us-east-1-1.iron.io`

ryantology committed:
Added reservaition id to delete method.

This is required in order to support the v3 API.
IronIO is rapidly moving to the v3 API as it has increased performance.
References Issue #10828

ryantology commented:
I have updated this to allow usage of the older version of the API. This should keep the patch backwards compatible. Because the version number of the API is a protected property (with no getter) I wasn't able to make a conditional in the unit test to test the different version of the API. Thoughts on how to update the unit tests?

GrahamCampbell commented:
Please send to 5.2, and bump the required ironmq package version.

ryantology commented:
This change should be backwards compatible (it is the only change that I needed to make in nodejs implementations). Due to **major** performance issues with push queues using the v1 API, an application I have built against 5.1 requires this minor change. I would rather not have myself and other developers introduce potential breaking changes into applications simply to support a change that can be back ported. I realize this is not a specific security or bug fix but the LTS version could really benefit from this patch IMHO.

One of the more confusing parts about installing IronMQ is that the v3 API (v4 branch) specifically says it is made for laravel 5.1 https://github.com/iron-io/iron_mq_php#branches

I guess if you feel strongly about this moving to 5.2 then I can continue to maintain my fork until 5.2 comes out.

GrahamCampbell commented:
Ping @taylorotwell.

GrahamCampbell commented:
You need to update the suggested dependencies we have then @ryantology.

ryantology commented:
I updated the suggested dependencies. Do you think there should be a comment here: https://github.com/laravel/laravel/blob/master/config/queue.php#L62 about the change in the endpoint. Because the change is so minor I overlooked it for a good amount of time trying to debug after upgrading. Or maybe it should be updated for people doing fresh installs.

GrahamCampbell commented on composer.json:
```diff
@@ -75,7 +75,7 @@
     "require-dev": {
         "aws/aws-sdk-php": "~3.0",
-        "iron-io/iron_mq": "~2.0",
+        "iron-io/iron_mq": "~2.0 || ~4.0",
```
please use `~2.0|~4.0`, and also update the other 2 places where this occurs too

ryantology replied:
Are you referring to the space between the `||` or that there should only be a single `|` I am really surprised the single `|` works :) The documentation for composer says to use double pipe https://getcomposer.org/doc/articles/versions.md#range

(updating to single pipe but just curious your feedback)

GrahamCampbell replied:
We always use the `|` syntax. See our master branch. We'd like to make sure stuff works on older versions of composer.

GrahamCampbell commented on src/Illuminate/Queue/IronQueue.php:
```diff
@@ -145,11 +145,17 @@ public function pop($queue = null)
      *
      * @param  string  $queue
      * @param  string  $id
+     * @param  string  $reservation_id Required for version 3
```
string|null

GrahamCampbell replied:
don't include the comment too

GrahamCampbell replied:
also, MUST be reservationId

GrahamCampbell replied:
maybe just `reservation` woild be fine tbh

GrahamCampbell commented on tests/Queue/QueueIronJobTest.php:
```diff
@@ -21,7 +21,7 @@ public function testFireProperlyCallsTheJobHandler()
     {
         $job = $this->getJob();
-        $job->getIron()->shouldReceive('deleteMessage')->once()->with('default', 1);
+        $job->getIron()->shouldReceive('deleteMessage')->once()->with('default', 1, null);
```
please don't update this test

GrahamCampbell replied:
it's important that this test passes without changing it to prove the changes are non-breaking

ryantology committed:
Updating composer and adjust method call

Conditional changed so that no parameter is passed in order to ensure
compatability with tests.
Will rebase/squish out after review.

ryantology commented:
I fixed the test. What I would like is to have a conditional test that actually verifies the `reservation_id` change. The issue is that the api/branch version info in IronMQ is protected. Do you know of a way to check the version or do you recommend another way to test? Going through composer seems really dirty but possible?

taylorotwell commented on src/Illuminate/Queue/Jobs/IronJob.php:
```diff
@@ -83,7 +83,11 @@ public function delete()
 
-        $this->iron->deleteMessage($this->getQueue(), $this->job->id);
+        if (property_exists($this->job, 'reservation_id')) {
+            $this->iron->deleteMessage($this->getQueue(), $this->job->id, $reservation_id);
```
Where is this $reservation_id variable set?

GrahamCampbell replied:
I think this is meant to say `$this->job->reservation_id` maybe?

GrahamCampbell commented on composer.json:
```diff
@@ -103,7 +103,7 @@
         "fzaninotto/faker": "Required to use the eloquent factory builder (~1.4).",
         "guzzlehttp/guzzle": "Required to use the Mailgun and Mandrill mail drivers (~5.3|~6.0).",
-        "iron-io/iron_mq": "Required to use the iron queue driver (~2.0).",
+        "iron-io/iron_mq": "Required to use the iron queue driver (~2.0|~4.0).",
```
Please update this in the other composer.json file for the queue component too.

ryantology replied:
Fixed in 44217c15731b10ba1b1b73f06a5ce09e69c96065 Sorry, didn't realize there was another composer file for queues.

ryantology committed:
Fixed reservation_id property and queue composer

taylorotwell commented:
Why continue supporting old verisons of the library at all if it's not a breaking change?

ryantology commented:
The v1 and v3 APIs run on different clusters. So if you already have an app running with queues in IronIO on v1 then you need to re-create the queues in v3 (queues are auto created though). It's a bit of a transition phase. You can use either one but if you were using the old one you need to change your configs to point to the new v3 endpoints. Does that make sense?

Essentially you can't run them in parallel. Your app can't have 1/2 servers pointing to v1 and other 1/2 at v3 as the queues are on different clusters.

taylorotwell commented:
We should just support the newest API. We will document it in the upgrade notes for people that are using the old one.

taylorotwell commented:
Can you rebase this. Will allow this in 5.1 since it's not bad and support only 4.0 if IronMQ in 5.2.

ryantology commented:
You turbo merged. Was just about to rebase :)

I'll make a separate PR against 5.2 that forces v4 library?

taylorotwell commented:
I already did it on master :) Thanks

On Fri, Nov 13, 2015 at 3:36 PM, Ryan White <notifications@github.com>
wrote:

> You turbo merged. Was just about to rebase :)
>
> I'll make a separate PR against 5.2 that forces v4 library?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/laravel/framework/pull/10848#issuecomment-156562582>.
>
